optax/_src/complex_valued.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Complex-valued optimization."""
+
+from dataclasses import dataclass
+from typing import NamedTuple
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from optax._src import base
+
+
+@jax.tree_util.register_pytree_node_class
+@dataclass
+class RealPair:
+  """A pair of real arrays split from a complex array."""
+  real: chex.Array
```

Comment:
No, I think it should be `chex.Array`. I would only split a complex array into a pair of real ones at the level of leaf nodes, not the whole pytree.

Reply:
Hmmm, not sure! I agree that we would only ever use it in a situation where the children are leaves so I agree this calls for `chex.Array`, but some jax internals might expect it to be a container that doesn't care about the type of its children (and in particular, whether they are leaves or trees). An example of this is https://github.com/google/jax/pull/7551. Here jax used to unflatten a tree with `None` values as children.

I think I'd still vote for using `chex.Array`, especially since we'd get problems with line 49 (`x.real + x.imaginary * 1j`) if `x.real` and `x.imaginary` are trees rather than arrays.